airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionHandler.kt

```diff
@@ -0,0 +1,175 @@
+
+const val COMMON_EXCEPTION_MESSAGE_TEMPLATE: String =
+    "Could not connect with provided configuration. Error: %s"
+
+const val DATABASE_CONNECTION_ERROR: String =
+    "Encountered an error while connecting to the database error"
+
+enum class FailureType {
+    CONFIG,
+    TRANSIENT
+}
+
+data class ConnectorErrorProfile(
+    var errorClass: String,
+    var regexMatchingPattern: String,
+    var failureType: FailureType,
+    var externalMessage: String,
+    var sampleInternalMessage: String,
+    var referenceLink: List<String>
+)
+
+/**
+ * This abstract class defines interfaces that will be implemented by individual connectors for
+ * translating internal exception error messages to external user-friendly error messages.
+ */
+open class ConnectorExceptionHandler {
+    @kotlin.jvm.JvmField
+    var DATABASE_READ_ERROR: String = "Encountered an error while reading the database"
+
+    protected open var connectorErrorDictionary: MutableList<ConnectorErrorProfile> =
+        mutableListOf()
+
+    /**
+     * Handles exceptions thrown by the connector. This method is the main entrance for handling
+     * exceptions thrown by the connector. It checks if the exception is a known exception, and if
+     * so, it emits the appropriate trace and external user-friendly error message. If the exception
+     * is not known, it rethrows the exception, which becomes a system error.
+     */
+    fun handleException(
+        e: Throwable,
+        cmd: Command,
+        outputRecordCollector: Consumer<AirbyteMessage>
+    ) {
+        ApmTraceUtils.addExceptionToTrace(e)
+        val rootException: Throwable? = getRootException(e)
+        val externalMessage: String? = getExternalMessage(rootException)
+        // error messages generated during check() needs special handling
+        if (cmd == Command.CHECK) {
+            outputRecordCollector.accept(
+                AirbyteMessage()
+                    .withType(AirbyteMessage.Type.CONNECTION_STATUS)
+                    .withConnectionStatus(
+                        AirbyteConnectionStatus()
+                            .withStatus(AirbyteConnectionStatus.Status.FAILED)
+                            .withMessage(externalMessage),
+                    ),
+            )
+        } else {
+            if (checkErrorType(rootException, FailureType.CONFIG)) {
+                AirbyteTraceMessageUtility.emitConfigErrorTrace(e, externalMessage)
+                exitProcess(1)
+            } else if (checkErrorType(rootException, FailureType.TRANSIENT)) {
+                AirbyteTraceMessageUtility.emitTransientErrorTrace(e, externalMessage)
+                exitProcess(1)
+            }
+            throw e
+        }
+    }
+
+    /**
+     * Initializes the error dictionary for the connector. This method shall include all the errors
+     * that are shared by all connectors. For instance, Debezium errors
+     */
+    open fun initializeErrorDictionary() {}
```

Comment:
if you don't want to make this an interface, why not replace this method with a class constructor?

Reply:
initially I tried to put this function inside a constructor.. then jetbrain gave me warning, suggesting it is not a good practice to do so in the constructor when the object is not fully initialized... which is a false positive...but I still followed its suggestion ;-)

But, now you helped me discover the init block! This is perfect for that, I will move the initialization function call into the init block of the subclasses and the parent handler class.

Reply:
> then jetbrain gave me warning, suggesting it is not a good practice to do so in the constructor when the object is not fully initialized...

The compiler gets anxious because it doesn't like calling non-private methods in the constructor because these might be overloaded by a subclass, but at the point in time when the constructor of the superclass is called the subclass hasn't been initialized yet (please correct me if I'm wrong @stephane-airbyte ). It's just a recipe for confusion.

In java, `javac` actually warns about this too. It's more sneaky in java because the `open` is implied there, unlike in kotlin.

Reply:
yes, now they are called inside constructor, and no longer need to be separately called after construction. Please take another look :-)